cpp/src/interop/arrow_to_cudf.cu

```diff
@@ -0,0 +1,297 @@
+    case arrow::Type::type::DATE64: return data_type(type_id::TIMESTAMP_MILLISECONDS);
+    case arrow::Type::type::STRING: return data_type(type_id::STRING);
+    case arrow::Type::type::DICTIONARY: return data_type(type_id::DICTIONARY32);
+    case arrow::Type::type::LIST: return data_type(type_id::LIST);
+    default: CUDF_FAIL("Unsupported type_id conversion to cudf");
+  }
+}
+
+namespace {
+struct dispatch_to_cudf_column {
+  template <typename T>
+  std::pair<std::unique_ptr<column>, column_view> operator()(arrow::Array const& array,
+                                                             data_type type,
+                                                             bool skip_mask,
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             cudaStream_t stream)
+  {
+    auto data_buffer = array.data()->buffers[1];
+    auto num_rows    = data_buffer->size() / sizeof(T);
+    auto has_nulls   = skip_mask ? false : array.null_bitmap_data() != nullptr;
+    auto col         = make_fixed_width_column(
+      type, num_rows, has_nulls ? mask_state::UNINITIALIZED : mask_state::UNALLOCATED, stream, mr);
+    auto mutable_column_view = col->mutable_view();
+    CUDA_TRY(cudaMemcpyAsync(mutable_column_view.data<void*>(),
+                             data_buffer->data(),
+                             data_buffer->size(),
+                             cudaMemcpyHostToDevice,
+                             stream));
+    if (has_nulls) {
+      CUDA_TRY(cudaMemcpyAsync(mutable_column_view.null_mask(),
+                               array.null_bitmap_data(),
+                               array.null_bitmap()->size(),
+                               cudaMemcpyHostToDevice,
+                               stream));
+    }
+
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(array.offset()),
+                          static_cast<size_type>(array.offset() + array.length())));
+  }
+
+  std::unique_ptr<rmm::device_buffer> get_mask_buffer(arrow::Array const& array,
+                                                      rmm::mr::device_memory_resource* mr,
+                                                      cudaStream_t stream)
+  {
+    return [&]() {
+      if (array.null_bitmap_data() == nullptr) {
+        return std::make_unique<rmm::device_buffer>(0, stream, mr);
+      } else {
+        auto mask = std::make_unique<rmm::device_buffer>(array.null_bitmap()->size(), stream, mr);
+        CUDA_TRY(cudaMemcpyAsync(mask->data(),
+                                 array.null_bitmap_data(),
+                                 array.null_bitmap()->size(),
+                                 cudaMemcpyHostToDevice,
+                                 stream));
+        return mask;
+      }
+    }();
```

Comment:
Is there a reason this is wrapped in an immediately-executed function expression?

Reply:
Just review comment to improvise.

Reply:
https://github.com/rapidsai/cudf/pull/5778#discussion_r462562920

Reply:
There's a "Better Code" recording on this anti-pattern. :)

Reply:
I think this is the pattern which was discussed to be better than if else cond, if I am not wrong.

Reply:
Please, just no.

Reply:
@rgsl888prabhu if I understand correctly, the original code implemented the "define then modify" anti-pattern with the null mask:
```c++
auto mask = rmm:device_buffer(0);
if (array.null_bitmap_data()) { mask = rmm::device_buffer(array.null_bitmap()); }
return std::make_unique(mask);
```

But the current version with two return statements avoids that anti-pattern by returning early, and the IIFE isn't adding anything. Am I misunderstanding something fundamental that the lambda is enabling?

Reply:
No, I think its just same nothing different.